### PR TITLE
Implement OAuthCallbackServlet to retrieve access tokens for Sheets API

### PR DESCRIPTION
**Pull request by cxfloeder**
The OAuthCallbackServlet takes in a authorization code from the Login Servlet. It makes a request to get the access tokens for the Sheets API. The access token is then parsed from the token request's response. The access token is then displayed on a webpage. I created a OAuthConstants class to hold common final variables between the Login and OAuthCallback Servlets.

### CODE REVIEW

File: src/main/java/data/SheetsServiceUtil.java [hunk 1]
```diff
@@ -32,10 +32,7 @@
 public class SheetsServiceUtil {
 
   public static Sheets getSheetsService() throws IOException, GeneralSecurityException {
-    GoogleCredential credential = SheetsAuthUtil.authorize();
-
-    return new Sheets.Builder(new UrlFetchTransport(), new JacksonFactory(), credential)
-      .setApplicationName("App Engine default service account")
-      .build();
+    // Create a spreadsheet by using the OAuth tokens.
```
Comment (ianjjohnson) [inline src/main/java/data/SheetsServiceUtil.java:35]: Is this comment meant to be a TODO?

File: src/main/java/servlets/LoginServlet.java [hunk 2]
```diff
@@ -24,19 +24,16 @@
...
 
 @WebServlet("/Login")
 public class LoginServlet extends HttpServlet {
   private static final Logger LOGGER = Logger.getLogger(LoginServlet.class.getName());
   private static final String CONTENT_TYPE_TEXT_HTML = "text/html;";
   private static final String REDIRECT_LINK = "/Login";
-  private final String CLIENT_ID = "client_id=150737768611-svndjtlklolq53g4ass4r3sqal2i31p5.apps.googleusercontent.com";
   private final String SCOPE = "scope=https://www.googleapis.com/auth/spreadsheets";
-  private final String REDIRECT_URI = "redirect_uri=";
   private final String OAUTH_LOGIN_URI = "https://accounts.google.com/o/oauth2/v2/auth";
```
Comment (ianjjohnson) [inline src/main/java/servlets/LoginServlet.java:35]: Probably this also belongs in OAuthConstants?

File: src/main/java/servlets/OAuthCallbackServlet.java [hunk 3]
```diff
@@ -30,25 +31,22 @@
...
-  private final String CODE = "https://accounts.google.com/o/oauth2/v2/auth";
   private final String TOKEN_URI = "https://oauth2.googleapis.com/token";
-
-  // Figure out a way to secure method to retrieve the client secret.
-  private final String CLIENT_SECRET = "";
-  private final String REDIRECT_URI = "redirect_uri= SOME CALLBACK URI";
-
+  private final String AUTH_CODE = "code=";
+  private final String CLIENT_SECRET = "client_secret=";
+  private final String SECRET_FILEPATH = "../../src/main/resources/secret.txt";
```
Comment (ianjjohnson) [inline src/main/java/servlets/OAuthCallbackServlet.java:44]: this one too?

File: src/main/java/data/SheetsServiceUtil.java (hunk 1 above)
Comment (cxfloeder) [reply src/main/java/data/SheetsServiceUtil.java:35]: Yes, I plan on using this class in the future.

File: src/main/java/servlets/LoginServlet.java [hunk 4]
```diff
@@ -53,17 +47,18 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   /** Build the OAuth consent page redirect URL. */
   private String getOAuthRedirectURL() {
-    String redirectUri = REDIRECT_URI + getRedirectUri();
-    
-    return String.format("%s?%s&%s&%s&%s", OAUTH_LOGIN_URI, CLIENT_ID,
-      RESPONSE_TYPE, redirectUri, SCOPE); 
+    return String.format("%s?%s&%s&%s&%s", OAuthConstants.OAUTH_LOGIN_URI, 
+      OAuthConstants.CLIENT_ID, OAuthConstants.RESPONSE_TYPE, getRedirectUri(), 
+      OAuthConstants.SCOPE);
```
Comment (cadecrandall) [inline src/main/java/servlets/LoginServlet.java:52]: can you put an example in a comment of what this string will look like?

File: src/main/java/servlets/LoginServlet.java [hunk 5]
```diff
@@ -53,17 +47,18 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      OAuthConstants.CLIENT_ID, OAuthConstants.RESPONSE_TYPE, getRedirectUri(), 
+      OAuthConstants.SCOPE); 
   }
 
   // Build a valid redirect URI to the OAuthCallbackServlet.
   private String getRedirectUri() {
     try {
-      URI domainUri = URI.create(System.getenv().get(ENVIRONMENT_VARIABLE));
-      return domainUri.resolve(OAUTH_CALLBACK_SERVLET).toString();
+      URI domainUri = URI.create(System.getenv().get(
```
Comment (frostyclawz) [inline src/main/java/servlets/LoginServlet.java:58]: Is this to get things for the URI? Just a little confusion.

Comment (cxfloeder) [reply src/main/java/servlets/LoginServlet.java:58]: This gets the redirect URI that is saved in the environment variable.

Comment (frostyclawz) [reply src/main/java/servlets/LoginServlet.java:58]: Thanks for the clarifying comment!

File: src/main/java/servlets/OAuthCallbackServlet.java [hunk 6]
```diff
@@ -30,25 +31,18 @@
...
-  // Figure out a way to secure method to retrieve the client secret.
-  private final String CLIENT_SECRET = "";
-  private final String REDIRECT_URI = "redirect_uri= SOME CALLBACK URI";
-
+  private static final Logger LOGGER = Logger.getLogger(LoginServlet.class.getName());
+  private final String SECRET_FILEPATH = "../../src/main/resources/secret.txt";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String error = request.getParameter("error");
```
Comment (jakecarlson1) [inline src/main/java/servlets/OAuthCallbackServlet.java:44]: Consider moving "error" to a constant.

File: src/main/java/servlets/OAuthCallbackServlet.java [hunk 7]
```diff
@@ -30,25 +31,18 @@
...
-    String code = request.getParameter("code");
+    String authCode = request.getParameter("code");
```
Comment (jakecarlson1) [inline src/main/java/servlets/OAuthCallbackServlet.java:45]: Consider moving "code" to a constant.

File: src/main/java/servlets/OAuthCallbackServlet.java [hunk 8]
```diff
@@ -57,29 +51,63 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
 
-    // Request the tokens.
+    // Request the access tokens.
     HttpClient httpClient = HttpClient.newHttpClient();
-    HttpRequest tokenRequest = HttpRequest.newBuilder(URI.create(TOKEN_URI))
-      .header("Content-Type", "application/x-www-form-urlencoded")
-      .POST(BodyPublishers.ofString(tokenRequestBody)).build();
+    HttpRequest tokenRequest = HttpRequest.newBuilder(URI.create(
+      OAuthConstants.TOKEN_URI)).header("Content-Type", 
+      "application/x-www-form-urlencoded").POST(BodyPublishers.ofString(
```
Comment (jakecarlson1) [inline src/main/java/servlets/OAuthCallbackServlet.java:68]: Consider moving this content type string to a constant.

Comment (jakecarlson1) [inline src/main/java/servlets/OAuthCallbackServlet.java:68]: This method would be a little easier to parse through if this was split into a few functions. Maybe move this to a "fetchAccessTokens" or "buildAccessTokensRequest" method?

File: src/main/java/servlets/OAuthCallbackServlet.java [hunk 9]
```diff
@@ -57,29 +51,63 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      BodyHandlers.ofString()).join();
     String tokenResponseBody = tokenResponse.body().toString();
 
     // Parses to find access token.
-    JsonObject tokenResponseObj = JsonParser.parseString(tokenResponseBody).getAsJsonObject();
+    JsonObject tokenResponseObj = JsonParser.parseString(tokenResponseBody)
+      .getAsJsonObject();
     JsonElement accessToken = tokenResponseObj.get("access_token");
 
     response.setContentType("text/html");
```
Comment (jakecarlson1) [inline src/main/java/servlets/OAuthCallbackServlet.java:80]: Consider making this a string constant.

File: src/main/java/servlets/LoginServlet.java (hunk 4 above)
Comment (jakecarlson1) [reply src/main/java/servlets/LoginServlet.java:52]: +1 for an example of what this string will look like.

File: src/main/java/data/SheetsServiceUtil.java (hunk 1 above)
Comment (jakecarlson1) [reply src/main/java/data/SheetsServiceUtil.java:35]: TODO comments should be structured like:
`// TODO(cxfloeder): Create a spreadsheet by using the OAuth tokens.`

File: src/main/java/servlets/OAuthCallbackServlet.java (hunk 8 above)
Comment (cxfloeder) [reply src/main/java/servlets/OAuthCallbackServlet.java]: I created a buildTokenRequest() and  parseAccessToken() method. That cleaned up the doGet() a little bit.

**Pull request merged**